### ISSUE TITLE
Compiler complains with error CS0266 that cannot convert 'FooLib.IFoo<T>' to 'FooLib.IFoo<T>'.  An explicit cast exists (are you missing a cast?).  Adding an explicit cast is not accepted by the IDE and code still does not build anyway.

### ISSUE THREAD

**Issue by astrohart** [dotnet/roslyn#51302]
**Version Used**: VS 2019 Enterprise 16.8.5

**Steps to Reproduce**:

1. Create an interface `IFoo<T>` in a class library in .NET Framework 4.8.
2. In another interface, `IBar<T>` in another class library in .NET Framework 4.8, same solution.
3. Create appropriate project references.
4. In interface `IBar<T>` create a method `IBar<T> HavingFoo(IFoo<T> foo)`
5. Derive a class from `IBar<T>` and implement the method thus (assume all the proper `using` statements are in place):

```
public class MyClass<T> : IBar<T>
{
    /* field */
    private IFoo<T> _foo;

    public IBar<T> HavingFoo<T>(IFoo<T> foo)
    {
        _foo = foo;    // error CS0266 (WTF?? Why??)
    }
    /* ... */
}
```
**Expected Behavior**:

Line `_foo = foo` compiles.

**Actual Behavior**:
`error CS0266: Cannot implicitly convert type 'FooLib.IFoo<T>' to 'FooLib.IFoo<T>'. An explicit conversion exists (are you missing a cast?)`

When you try to do a `CTRL+.` and add an explicit cast, the IDE does not add the explicit cast.  Adding an explicit cast by hand, i.e., to make the offending line read `_foo = (IFoo<T>)foo;` makes the IDE "dim" the cast, and raise `IDE0004` saying cast is unnecessary.

This is bizarro.  

It works just fine and compiles without errors if I dependency-inject in the constructor:

```
public class MyClass<T> : IBar<T>
{
    /* field */
    private readonly IFoo<T> _foo;

    public MyClass(IFoo<T> foo)
    {
        _foo = foo;    // compiles fine
    }
    /* ... */
}
```

That is just plain weird: First of all, `CS0266` is telling me the compiler cannot convert a class to itself (?? huh ??) and then it will compile dependency-injection just fine but not the fluent method.

What is going on?  I am wondering if this is even reproducible. I am just plain flummoxed.  I have deep experience coding in C#, so I think I know what I am doing (i.e., this is not some "n00b mistake").

**What I've Tried**
File > Reload Workspace -> did not work; Close VS, delete `bin` and `obj` folders, re open VS, do a `Rebuild Solution` -> did not work.

**Comment by jaredpar** [comment]
Based on the error here I'm guessing the underlying issue is there are two types named `FooLib.IFoo<T>`. Can you post a full repro here that we can take a look at?

**Comment by astrohart** [comment]
Hi-

Well, I went back and double-checked the [Microsoft documentation on Generic methods](https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/generics/generic-methods).  It has the following example, which most closely matched my particular Roslyn use case:
```
class SampleClass<T>
{
    void Swap(ref T lhs, ref T rhs) { }
}
```
As can be seen in the OP, I want to utilize the same type parameter as the containing class, in the method parameters of the `HavingFoo` method.  It was extremely late at night when I filed this post, and I forgot to mention that the signature of my `HavingFoo` method that was erroring was:
```
public IBar<T> HavingFoo<T>(IFoo<T> foo)
{
   /* ... */
}
```
Upon reading [the Microsoft docs](https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/generics/generic-methods) -- and sleeping for a full 8 hours -- I stumbled upon the notion that the `<T>` in `HavingFoo<T>` might be confusing the compiler.  

After all, it was warning me about it -- I forget the warning number exactly.  (I only read the docs after having gotten sleep -- you know how it is when you bang your head against the wall over and over fruitlessly and it's really just your brain saying, "Help!  I need sleep!" ;-p

The [docs](https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/generics/generic-methods) made me realize that the `<T>` in the method signature was not the same `T` (which I was intending) coming down from the containing class.  Removing the `<T>` from the method signature did the trick and solved the error.

I am a newbie to generic methods.  The need for this fluent method just to initialize a dependency comes from the fact that I am using a factory to create instances of the containing class and another factory to provide the dependency.  Therefore I am needing a fluent method to initialize my dependency, so I can do the injection outside of the factory.

I can't hand the dependency to the factory because then one of my other class libraries would depend too heavily on the third-party libraries I am extending my software with.

Making my code read:
```
public class MyClass<T> : IBar<T>
{
    /* field */
    private IFoo<T> _foo;

    public IBar<T> HavingFoo(IFoo<T> foo)
    {
        _foo = foo;    // compiles
    }
    /* ... */
}
```
solved it.
